File: services/inventory/src/modules/inventory/infrastructure/messaging/outbox-publisher.worker.ts
import { Inject, Injectable, Logger } from '@nestjs/common';
import { Interval } from '@nestjs/schedule';
import {
  OUTBOX_REPOSITORY,
  STOCK_EVENT_PUBLISHER,
} from '../../application/ports/stock.tokens';
import {
  OutboxRepository,
  StockEventPublisher,
} from '../../application/ports/stock.repositories';

@Injectable()
export class OutboxPublisherWorker {
  private readonly logger = new Logger(OutboxPublisherWorker.name);
  private running = false;

  constructor(
    @Inject(OUTBOX_REPOSITORY)
    private readonly outboxRepository: OutboxRepository,
    @Inject(STOCK_EVENT_PUBLISHER)
    private readonly publisher: StockEventPublisher,
  ) {}

  @Interval(1000)
  async publishPending(): Promise<void> {
    if (this.running) {
      return;
    }

    this.running = true;
    try {
      const events = await this.outboxRepository.findPending(50);
      for (const record of events) {
        try {
          await this.publisher.publish({
            eventId: record.eventId,
            eventType: record.eventType as
              'inventory.stock_added' | 'inventory.stock_debited',
            aggregateId: record.aggregateId,
            occurredAt: record.occurredAt,
            version: record.version,
            payload: record.payload,
          });
          await this.outboxRepository.markPublished(record.eventId);
        } catch (error) {
          await this.outboxRepository.markFailed(
            record.eventId,
            error instanceof Error ? error.message : String(error),
          );
          this.logger.warn(
            'Evento de Estoque mantido na Outbox para retry: ' + record.eventId,
          );
        }
      }
    } finally {
      this.running = false;
    }
  }
}

File: services/inventory/src/main/java/br/com/delivery/inventory/infrastructure/messaging/OutboxPublisherWorker.java
package br.com.delivery.inventory.infrastructure.messaging;

import br.com.delivery.inventory.application.ports.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

@Component
public class OutboxPublisherWorker {
  private static final Logger log = LoggerFactory.getLogger(OutboxPublisherWorker.class);
  private final OutboxRepository outbox;
  private final StockEventPublisher publisher;

  public OutboxPublisherWorker(OutboxRepository outbox, StockEventPublisher publisher) {
    this.outbox = outbox;
    this.publisher = publisher;
  }

  @Scheduled(fixedDelayString = "${inventory.outbox.delay:1000}")
  public void publishPending() {
    for (var event : outbox.findPending(50)) {
      try {
        publisher.publish(event);
        outbox.markPublished(event.eventId());
      } catch (RuntimeException e) {
        outbox.markFailed(event.eventId(), "Publicação não confirmada; aguardando nova tentativa");
        log.warn("Evento {} mantido na Outbox", event.eventId());
      }
    }
  }
}
